File: backend/src/modules/topos/topos.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, Like, LessThan } from 'typeorm';
import { ToposD } from './entities/topos-d.entity';
import { PaginationDto, PaginatedResult } from '../../common/dto/pagination.dto';

@Injectable()
export class ToposService {
  constructor(
    @InjectRepository(ToposD)
    private toposDRepository: Repository<ToposD>,
  ) {}

  async findAllDialogs(dto: PaginationDto): Promise<PaginatedResult<ToposD>> {
    const { page, limit, keyword } = dto;
    const skip = (page - 1) * limit;
    const where: any = {};
    if (keyword) where.a_callid = Like(`%${keyword}%`);
    const [items, total] = await this.toposDRepository.findAndCount({
      where, skip, take: limit, order: { rectime: 'DESC' },
    });
    return new PaginatedResult(items, total, page, limit);
  }

  async cleanup(beforeDate: Date): Promise<number> {
    const result = await this.toposDRepository.delete({
      rectime: LessThan(beforeDate),
    });
    return result.affected || 0;
  }

  async getStats(): Promise<any> {
    const total = await this.toposDRepository.count();
    return { total };
  }
}

File: kam-admin-module-trace/src/main/java/io/kamailio/admin/modules/topos/service/ToposService.java
package io.kamailio.admin.modules.topos.service;

import io.kamailio.admin.common.dto.PaginatedResult;
import io.kamailio.admin.common.dto.PaginationDto;
import io.kamailio.admin.modules.topos.entity.ToposD;
import io.kamailio.admin.modules.topos.repository.ToposDRepository;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.HashMap;
import java.util.Map;

@Service
public class ToposService {
    private final ToposDRepository repository;

    public ToposService(ToposDRepository repository) {
        this.repository = repository;
    }

    public PaginatedResult<ToposD> findAllDialogs(PaginationDto dto) {
        var page = dto.getKeyword() != null && !dto.getKeyword().isBlank()
                ? repository.findByACallidContaining(dto.getKeyword(), PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by(Sort.Direction.DESC, "rectime")))
                : repository.findAll(PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by(Sort.Direction.DESC, "rectime")));
        return new PaginatedResult<>(page.getContent(), page.getTotalElements(), dto.pageOrDefault(), dto.limitOrDefault());
    }

    @Transactional
    public int cleanup(Instant beforeDate) {
        return repository.deleteByRectimeBefore(beforeDate);
    }

    public Map<String, Object> getStats() {
        Map<String, Object> result = new HashMap<>();
        result.put("total", repository.count());
        return result;
    }
}
